pkg/sdkserver/sdkserver.go

```diff
@@ -762,20 +790,202 @@ func (s *SDKServer) GetCounter(ctx context.Context, in *alpha.GetCounterRequest)
+	batchCounter, batchOk := s.gsCounterUpdates[name]
+
+	if !batchOk {
+		gs, err := s.gameServer()
+		if err != nil {
+			return nil, err
+		}
+
+		counter, ok := gs.Status.Counters[name]
+		// We didn't find the Counter named key in the gameserver.
+		if !ok {
+			return nil, errors.Errorf("NOT_FOUND. %s Counter not found", name)
+		}
+
+		batchCounter.counter = *counter.DeepCopy()
+	}
+
+	switch {
+	case in.CounterUpdateRequest.CountDiff != 0: // Update based on if Client call is CountIncrement or CountDecrement
+		count := batchCounter.counter.Count
+		if batchCounter.countSet != nil {
+			count = *batchCounter.countSet
+		}
+		count += batchCounter.diff + in.CounterUpdateRequest.CountDiff
+		// Verify that 0 <= Count >= Capacity
+		capacity := batchCounter.counter.Capacity
+		if batchCounter.capacitySet != nil {
+			capacity = *batchCounter.capacitySet
+		}
+		// Case where Count > Capacity we want to allow decrementation but not incrementation
+		if count < 0 || (in.CounterUpdateRequest.CountDiff > 0 && count > capacity) {
+			return nil, errors.Errorf("OUT_OF_RANGE. Count must be within range [0,Capacity]. Found Count: %d, Capacity: %d", count, capacity)
+		}
+		batchCounter.diff += in.CounterUpdateRequest.CountDiff
+	case in.CounterUpdateRequest.Count != nil: // Update based on if Client call is CountSet
+		// Verify that 0 <= Count >= Capacity
+		countSet := in.CounterUpdateRequest.Count.GetValue()
+		capacity := batchCounter.counter.Capacity
+		if batchCounter.capacitySet != nil {
+			capacity = *batchCounter.capacitySet
+		}
+		if countSet < 0 || countSet > capacity {
+			return nil, errors.Errorf("OUT_OF_RANGE. Count must be within range [0,Capacity]. Found Count: %d, Capacity: %d", countSet, capacity)
+		}
+		batchCounter.countSet = &countSet
+		// Clear any previous CountIncrement or CountDecrement requests, and add the CountSet as the first item.
+		batchCounter.diff = 0
+	case in.CounterUpdateRequest.Capacity != nil: // Updated based on if client call is CapacitySet
+		if in.CounterUpdateRequest.Capacity.GetValue() < 0 {
+			return nil, errors.Errorf("OUT_OF_RANGE. Capacity must be greater than or equal to 0. Found Capacity: %d", in.CounterUpdateRequest.Capacity.GetValue())
+		}
+		capacitySet := in.CounterUpdateRequest.Capacity.GetValue()
+		batchCounter.capacitySet = &capacitySet
+	default:
+		return nil, errors.Errorf("INVALID_ARGUMENT. Malformed CounterUpdateRequest: %v", in.CounterUpdateRequest)
+	}
+
+	s.gsCounterUpdates[name] = batchCounter
+
+	// Queue up the Update for later batch processing by updateCounters.
+	s.workerqueue.Enqueue(cache.ExplicitKey(updateCounters))
+	return &alpha.Counter{}, nil
+}
+
+// updateCounter updates the Counters in the GameServer's Status with the batched update requests.
+func (s *SDKServer) updateCounter(ctx context.Context) error {
+	if !runtime.FeatureEnabled(runtime.FeatureCountsAndLists) {
+		return errors.Errorf("%s not enabled", runtime.FeatureCountsAndLists)
```

Comment:
A general question about the error handling: will the workerqueue retry on this kind of error? I think it should not because it is not possible to retry successfully, but I'm not sure what's the current behavior.

Reply:
That is a good point - yes it will. So we should be sure that when we throw an error here it's because we want to retry.

Reply:
I can probably take this out -- the SDK should reject any of these errors in UpdateCounter before it gets put in the worker queue.

Reply:
Either that or log the error, and return nil, so it gets dropped immediately.